### PR TITLE
Add new functionality for returning an x509 file

### PR DESCRIPTION
**Pull request by nick-child-ibm**
In this commit, I ported Ken Goldman's x509 generator code from 2013. This will allow for a client to request an x509 version of their public key. The x509 can be either self signed or signed from a CA, depending on priviledges. The system administrator can make a CA by generating a new key with an auxiliary file containing signature/digest algorithms, validity, final and Issuer fields. The significance of these fields can be found in Kens documentation. The key to be signed in the ceritficate can be a any project if the certifying key is declared as final. If not, then an auxiliary file will be needed for the key to be certified. This will contain all the same fields as the ceritifying aux file. If the certifying key is listed as final than this means that it is the final issuer in the certificate chain, it will be used to sign other projects only. The auxiliary file for these final issuers must also contain subject fields that describes the project it is signing. This is all described in kens documentation. To request an x509 from the server side, the user can run './sf_client ./sf_client -project <project name1> -comments "Testing sign cert" -param "-certify_project <project name2> -usr <userid>" -url sftp://<userid>@argus4.rch.stglabs.ibm.com -epwd <encrypted pswd> -pkey <rsa_private> ' where <project name1> is certifying/signing <project_name2> .
NOTE: You will likely need to increase the file_max field of your framework config, since x509s are larger than other outputed filed
This commit also comes with testing scripts. To run, follow the below steps:
1. run keygen commands described in generateTestFiles.sh
2. Add yourself to senders in the config files for issuer_final issuer_root code_signer
3. run generateTestFiles.sh
4. export FRAMEWORK_CONFIG_FILE to point to regression.cfg
5. run the framework
6. switch over to the client machine
7. fill in user specific fields in sign_cert_test.sh
8. run sign_cert_test.sh

Kens Documentation:
[sign_cert documentation](https://github.com/open-power/sb-signing-framework/files/5968809/sign_cert_pdf.pdf)

### CODE REVIEW

File: src/signframework/sign_cert/cca_functions.c
Comment (cjengel) [inline src/signframework/sign_cert/cca_functions.c:1]: The current design could use fixing but common routines like cca_function.c/h exist in the signframework directory and then the various tools just create a link to it .  So if there are changes in here they should be merged into the main copy and links inserted in sign_cert to ../cca_functions.c

**Review by mtvaught-ibm — CHANGES_REQUESTED** [review]
Only looked through the function code c/h files so far, comparing them to Ken's original code.

The "generation" of the regression test config files don't make sense to me. Why not just generate them once and them commit them under the test/ directory? Then they just exist and the tester can point FRAMEWORK_CONFIG_FILE to the testing config?

**Comment by mtvaught-ibm** [comment]
A handful of the files you pushed up have a  IBM Confidential license header on them. Please discuss with @cjengel and Ken to determine what the correct licensing of these files are and update the headers accordingly.

**Comment by nick-child-ibm** [comment]
> Only looked through the function code c/h files so far, comparing them to Ken's original code.
> 
> The "generation" of the regression test config files don't make sense to me. Why not just generate them once and them commit them under the test/ directory? Then they just exist and the tester can point FRAMEWORK_CONFIG_FILE to the testing config?

Well I was worried about committing static files that were only used for testing. Are you guys planning on shipping this project with test scripts/files (noticing now that no other features have testing data)? If no, then I can commit the the static files and you folks can remove them before merging. Otherwise, I would think it would make more sense from a tester perspective to see where/why/how these files were generated and I always think it makes more sense to generate static files rather than shipping them. Also, if we choose to push the test files, should I remove the generator script? Let me know your thoughts and I can make the appropriate fixes.

File: src/signframework/sign_cert/makefile
Comment (nick-child-ibm) [reply src/signframework/sign_cert/makefile:58]: If we decide to keep the test file generator script then this does the cleanup for it.

Comment (mtvaught-ibm) [reply src/signframework/sign_cert/makefile:57]: My point is that the the test file generator script has nothing to do with building the project. So it is strange to have the cleanup for that script as part of the makefile.

**Comment by mtvaught-ibm** [comment]
> > Only looked through the function code c/h files so far, comparing them to Ken's original code.
> > The "generation" of the regression test config files don't make sense to me. Why not just generate them once and them commit them under the test/ directory? Then they just exist and the tester can point FRAMEWORK_CONFIG_FILE to the testing config?
> 
> Well I was worried about committing static files that were only used for testing. Are you guys planning on shipping this project with test scripts/files (noticing now that no other features have testing data)? If no, then I can commit the the static files and you folks can remove them before merging. Otherwise, I would think it would make more sense from a tester perspective to see where/why/how these files were generated and I always think it makes more sense to generate static files rather than shipping them. Also, if we choose to push the test files, should I remove the generator script? Let me know your thoughts and I can make the appropriate fixes.

This project doesn't get "shipped" in the conventional sense (or at least, not to my knowledge). On an install to a system the entire project is pulled and then compiled.

I'll let @cjengel weigh in on whether he sees an issue with having test files directly committed.

Perhaps a middle-ground would be to create a src/test directory to house the testing files and then have a script that could copy them out of src/test and into src/signframework. That would keep them out of the main framework on deployed systems, but also make it easier to see/edit the test files in the future.

**Comment by nick-child-ibm** [comment]
Sounds good, I have no problem doing either. Let me know what you guys think is best and I will make the adjustments. Then, I can fix the test cleanup.

File: src/client/sign_cert_test.sh
Comment (cjengel) [inline src/client/sign_cert_test.sh:8]: You missed a license header

**Review by cjengel — CHANGES_REQUESTED** [review]
Only issue I see is the one license header.  Need Matt to review as well.

File: src/signframework/sign_cert/tests/issuer_final_sample.cfg
Comment (mtvaught-ibm) [inline src/signframework/sign_cert/tests/issuer_final_sample.cfg:29]: You may want to remove your email

The other sample configs use joe@example.com

File: src/signframework/sign_cert/tests/issuer_final_sample.cfg
Comment (mtvaught-ibm) [inline src/signframework/sign_cert/tests/issuer_final_sample.cfg:39]: You may want to remove your email/id

The other sample configs use fred=fred@example.com

File: src/signframework/sign_cert/tests/issuer_final_sample_aux.cfg
Comment (mtvaught-ibm) [inline src/signframework/sign_cert/tests/issuer_final_sample_aux.cfg:10]: Remove IBM specific things

File: src/signframework/sign_cert/tests/issuer_final_sample_aux.cfg
Comment (mtvaught-ibm) [inline src/signframework/sign_cert/tests/issuer_final_sample_aux.cfg:20]: same here

File: src/signframework/sign_cert/tests/issuer_final_sample_aux.cfg
Comment (mtvaught-ibm) [inline src/signframework/sign_cert/tests/issuer_final_sample_aux.cfg:8]: Maybe pick a more "standard" validity range. I think the "standard" right now is a max of 2 years?

File: src/signframework/sign_cert/tests/issuer_root_sample.cfg
Comment (mtvaught-ibm) [inline src/signframework/sign_cert/tests/issuer_root_sample.cfg:40]: Change to non-identifying id/email

File: src/signframework/sign_cert/tests/issuer_root_sample.cfg
Comment (mtvaught-ibm) [inline src/signframework/sign_cert/tests/issuer_root_sample.cfg:31]: change to non-identifying email

File: src/signframework/sign_cert/tests/issuer_root_sample.cfg
Comment (mtvaught-ibm) [inline src/signframework/sign_cert/tests/issuer_root_sample.cfg:3]: remove

File: src/signframework/sign_cert/tests/issuer_root_sample_aux.cfg
Comment (mtvaught-ibm) [inline src/signframework/sign_cert/tests/issuer_root_sample_aux.cfg:3]: remove IBM information

File: src/signframework/sign_cert/tests/issuer_root_sample_aux.cfg
Comment (mtvaught-ibm) [inline src/signframework/sign_cert/tests/issuer_root_sample_aux.cfg:7]: max 2 years

File: src/signframework/sign_cert/tests/code_signer_sample.cfg
Comment (mtvaught-ibm) [inline src/signframework/sign_cert/tests/code_signer_sample.cfg:28]: email

File: src/signframework/sign_cert/tests/code_signer_sample.cfg
Comment (mtvaught-ibm) [inline src/signframework/sign_cert/tests/code_signer_sample.cfg:37]: email/id

File: src/signframework/sign_cert/tests/base_regression.cfg
Comment (mtvaught-ibm) [inline src/signframework/sign_cert/tests/base_regression.cfg:54]: remove IBM ldap information

File: src/signframework/sign_cert/tests/base_regression.cfg
Comment (mtvaught-ibm) [inline src/signframework/sign_cert/tests/base_regression.cfg:61]: change to something generic

**Review by mtvaught-ibm — CHANGES_REQUESTED** [review]
Need to remove IBM information, but other than that I think it is good to go.

File: src/signframework/sign_cert/tests/issuer_root_sample.cfg
Comment (mtvaught-ibm) [reply src/signframework/sign_cert/tests/issuer_root_sample.cfg:3]: Still need to remove this line

**Review by mtvaught-ibm — CHANGES_REQUESTED** [review]

**Comment by nick-child-ibm** [comment]
Let me know if there is anything else I can do.

**Comment by cjengel** [comment]
Looks good to me

**Review by mtvaught-ibm — APPROVED** [review]
@cjengel This PR should be merged using "squash and merge"

**Pull request merged**
